### PR TITLE
returning source error message instead of generic one

### PR DESCRIPTION
**Pull request by julsemaan**
# Description

Return the error message from the auth source instead of a generic 'wrong username/password' if the source provides one
# Impacts

login on the portal
# Delete branch after merge

YES
# NEWS file entries
## Enhancements
- Allow precise error messages from the authentication source when providing invalid credentials on the captive portal

### CODE REVIEW

**Comment by jrouzierinverse** [comment]
If there are multiple sources and all of them failed which error message should we use?

**Comment by julsemaan** [comment]
Right now, it uses the last one and I don't see how we could go around that as our circular search through all the sources will not always show the right message of the array

Coupling this with #839 will show the right error message though

If you have a good idea, I'm open for it

**Comment by jrouzierinverse** [comment]
The only idea I would have is to show them the error message.

What problem are you trying to solve with the pull request?

**Comment by julsemaan** [comment]
Well right now, when you're doing null auth and you enter an invalid e-mail, it uses the same generic message instead of using the one of the null source. Meaning an honnest mistake in an e-mail will show you 'Invalid username/password' which is not the case at all.

**Comment by jrouzierinverse** [comment]
The work I did in the branch feature/null-auth-non-exclusive will solve that problem by treating a null source like an external source

**Comment by julsemaan** [comment]
Doesn't solve the fact that we're using the same error message everywhere though.

Let's say there is an LDAP connection error. We should be showing that with along with a 'Please try again later' instead of showing 'Invalid username/password' when it was actually right.

We do have the info messages populated in authentication so why not leverage them to make the portal more user friendly...

**Comment by jrouzierinverse** [comment]
That make sense.
However at the moment each source have different error message for the same type of event.

All sources should return the same message for the same type of event.

Invalid credentials should say.

'Wrong username or password.'

When there is a connection problem.

'Please try again later'

**Comment by julsemaan** [comment]
Correct, this part is still to-do on my side before this hits devel

**Comment by julsemaan** [comment]
Messages reworked, re-review

File: lib/pf/authentication.pm [hunk 1]
```diff
@@ -255,8 +256,8 @@ sub authenticate {
         }
     }
 
-    $logger->trace("Authentication failed for '$username' for all ".scalar(@sources)." sources");
-    return ($FALSE, 'Wrong username or password.');
+    $logger->trace("Authentication failed for '$display_username' for all ".scalar(@sources)." sources");
+    return ($FALSE, $message ? $message : 'Wrong username or password.');
 }
```
Comment (jrouzierinverse) [inline lib/pf/authentication.pm:261]: All success/error messages should be in constants.

**Comment by julsemaan** [comment]
Re-review

File: lib/pf/constants/authentication.pm [hunk 2]
```diff
@@ -26,6 +26,15 @@ our @SOURCES = __PACKAGE__->sources();
...
 
 
+=head2 Messages
+
+Replies for authentication success/error
+
+=cut
+
+Readonly our $INVALID_CREDENTIALS => ;
+Readonly our $AUTH_SUCESS
```
Comment (jrouzierinverse) [inline lib/pf/constants/authentication.pm]: This looks incomplete

**Comment by jrouzierinverse** [comment]
where is the module for pf::constants::authentication::messages

**Comment by julsemaan** [comment]
Fixed

File: lib/pf/authentication.pm [hunk 3]
```diff
@@ -242,6 +242,7 @@ sub authenticate {
 
     $logger->debug(sub {"Authenticating '$username' from source(s) " . join( ', ', map { $_->id } @sources ) });
 
+    my $message;
     foreach my $current_source (@sources) {
         my ($result, $message);
```
Comment (jrouzierinverse) [inline lib/pf/authentication.pm]: This message overrides the one above

File: lib/pf/Authentication/Source/LDAPSource.pm [hunk 4]
```diff
@@ -93,15 +94,15 @@ sub authenticate {
...
   if ($result->is_error) {
     $logger->error("[$self->{'id'}] Unable to execute search $filter from $self->{'basedn'} on $LDAPServer:$LDAPServerPort");
-    return ($FALSE, 'Unable to validate credentials at the moment');
+    return ($FALSE, $COMMUNICATION_ERROR_MSG);
   }
 
   if ($result->count == 0) {
     $logger->warn("[$self->{'id'}] No entries found (". $result->count .") with filter $filter from $self->{'basedn'} on $LDAPServer:$LDAPServerPort");
-    return ($FALSE, 'Invalid login or password');
+    return ($FALSE, $COMMUNICATION_ERROR_MSG);
```
Comment (jrouzierinverse) [inline lib/pf/Authentication/Source/LDAPSource.pm]: This mean there are no user not a communication error

**Pull request merged**